### ISSUE TITLE
Bug: Copying Multiple Elements (ordering by)

### ISSUE THREAD

**Issue by 7flash** [excalidraw/excalidraw#6268]
It seems ordering happen by the time when element added, but expected behaviour to be selected by click number when doing ctrl+click either position when selecting with mouse area

In following example, expected to receive "1 2 3" but it shows "1 3 2" because I created "3" element first

![second22-first](https://user-images.githubusercontent.com/4569866/220543354-304984d7-f1f9-4f7a-8339-4101a23b76cf.gif)

**Comment by shuguang-lv** [comment]
Hello~ Is this issue still open? If so, can I work on this?

**Comment by shuguang-lv** [comment]
@ad1992 Hi~ Is this issue still open? If so, can I work on this too?

**Comment by dwelle** [comment]
Yup. Let's spec it out first though.

I'd not do selection order because that seems too cryptic / inconsistent, and doesn't solve box selection or select-all -> copy.

With spatial ordering, I'd do it as if reading a book: left to right

![image](https://github.com/excalidraw/excalidraw/assets/5153846/e2f191b2-b2c4-4f7d-b4da-72eeb25d3294)

More complex example:

![image](https://github.com/excalidraw/excalidraw/assets/5153846/7f819176-38fd-4da7-ba7c-e0502d367e5b)

**Comment by shuguang-lv** [comment]
@7flash @dwelle Hi, thanks for your explanation. From my perspective, the ordering of copied elements in the text should follow the order by which they are selected and respect active language settings. Is my understanding correct?

**Comment by shuguang-lv** [comment]
I can observe that currently, during copying as text, the ordering of selected elements follows the order by which they are created.

**Comment by dwelle** [comment]
@shuguang-lv hey. Actually, I'd implement the left-to-right, top-to-bottom approach as shown on the images. And disregard the active language for now.

**Comment by dwelle** [comment]
Thinking about it more, the way I described (or rather not described) the spatial algorithm is too simplistic and wouldn't work well with common cases where text elements that people would normally consider on a single line would not be considered on the same line by the algo because the elements would not have the same `y` position (unless you always aligned all elements vertically):

![image](https://github.com/excalidraw/excalidraw/assets/5153846/b2b90971-44ea-4347-96b3-256f90edcf95)

A better algo would be to take the top-most text element, compute its bounding box horizontal center line, and consider all elements whose bounding boxes touch that line to be on the same line with it. Then go left to right as normal. Afterwards, take the next top-most text element (of elements that are left) and do it again.

![image](https://github.com/excalidraw/excalidraw/assets/5153846/a0d87ab4-87a7-45ef-897e-961dbed42b4e)

I came up with just after a short amount of thinking. If anyone can suggest a better heuristic, feel free!

I asked GPT what it thinks, and it came up with a similar algo, though it's using tolerances instead:

![image](https://github.com/excalidraw/excalidraw/assets/5153846/51452b64-dc35-4549-8d1e-fd2f0f343060)

---

Another thing to consider is the output we should copy to clipboard. Right now, we copy each text shape on its own line. We could concat all "single line" shapes in a single line separated by spaces, but maybe that's assuming too much on the user's part. So let's start with keeping the current behavior (each shape its own paragraph), and go from user feedback afterwards.

**Comment by shuguang-lv** [comment]
@dwelle Hi, may I ask what's the base point of an element when calculating its xy coordinates? i.e. top-left corner, bottom-right corner, or the centroid?

**Comment by dwelle** [comment]
`element.x` and `element.y` reflects the top-left corner.

To get the element bounds (top-left, bottom-right), including (center-x, center-y), you can use the `getElementAbsoluteCoords()` helper from `src/element/bounds.ts`.

**Comment by shuguang-lv** [comment]
Thank you. That's a great hint to me!

**Comment by shuguang-lv** [comment]
@dwelle Hi~ Regarding your solution 2 above, I have a question. What if texts are in significantly different sizes? Like the case below, "1" touches the center lines of both "4" and "3". Which two should we think are on the same line?

<img width="132" alt="image" src="https://github.com/excalidraw/excalidraw/assets/56506969/3f9e53b2-edda-44fe-a3c2-861ae881b047">

**Comment by dwelle** [comment]
I'd do `4 1 3`, which conforms to the suggested algorithm.

With edge cases like this it's hard. I can image we could do `4 3 1`, but it'd be hard to write a generic algo for it.

**Comment by shuguang-lv** [comment]
Yes, I think so. I've submitted a pr with a naive solution. Could you review it to see if I'm in the right direction? Thanks
